src/Framework/Blocks/BlockCollection.php

```diff
@@ -79,24 +79,117 @@ public function getBlocks(): array
     /**
      * @unreleased
      *
-     * @param BlockModel[] $blocks
-     * @param string $name
-     * @return BlockModel|null
+     * @return BlockCollection|null
      */
-    public function findByName(string $name, array $blocks = null)
+    public function findByName(string $blockName, int $blockIndex = 0, string $return = 'self')
     {
-        if ($blocks === null) {
-            $blocks = $this->blocks;
+        return $this->findByNameRecursive($blockName, $blockIndex, $return);
+    }
+
+    /**
+     * @unreleased
+     *
+     * @return BlockModel|BlockCollection|null
+     */
+    private function findByNameRecursive(string $blockName, int $blockIndex = 0, string $return = 'self', BlockCollection $blockCollection = null, int &$count = 0)
+    {
+        if (!$blockCollection) {
+            $blockCollection = $this;
         }
 
-        foreach ($blocks as $block) {
-            if ($block->name === $name) {
-                return $block;
-            } else if ($block->innerBlocks) {
-                return $this->findByName($name, $block->innerBlocks->blocks);
+        foreach ($blockCollection->blocks as $block) {
+            if ($block->name === $blockName) {
+                $count++;
+
+                if ($count === $blockIndex + 1) {
+                    if ($return === 'self') {
+                        return $block;
+                    } elseif ($return === 'parent') {
+                        return $blockCollection;
+                    }
+                }
+            } elseif ($block->innerBlocks) {
+                $result = $this->findByNameRecursive($blockName, $blockIndex, $return, $block->innerBlocks, $count);
+                if ($result) {
+                    return $result;
+                }
             }
         }
 
         return null;
     }
+
+    /**
+     * @unreleased
+     */
+    public function insertBefore(string $blockName, BlockModel $block, int $blockIndex = 0): BlockCollection
+    {
+        $blockCollection = $this->findByName($blockName, $blockIndex, 'parent');
```

Comment:
I think we should use the internal find method for these methods so any changes to the public one will not affect these ones.

Reply:
Sure, no problem.